File: backend_java_spring_boot/myPizza/src/main/java/backendapp/myPizza/services/MenuService.java
package backendapp.myPizza.services;

import backendapp.myPizza.Models.entities.Menu;
import backendapp.myPizza.Models.entities.Product;
import backendapp.myPizza.Models.resDTO.ConfirmRes;
import backendapp.myPizza.exceptions.BadRequestException;
import backendapp.myPizza.repositories.MenuRepository;
import lombok.Setter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Service
public class MenuService {
    @Autowired
    private MenuRepository menuRp;

    public Page<Menu> getMenu(Pageable pageable) {
        return menuRp.findAll(pageable).map(m -> {
            if (m.getItem() instanceof Product) ((Product) m.getItem()).setProductTotalAmount();
            return m;
        });
    }

    public ConfirmRes saveMenu(List<UUID> menuIds) throws BadRequestException {
        List<Menu> newMenuList = new ArrayList<>();
        for (UUID menuId : menuIds) {
            Menu oldMenu = menuRp.findById(menuId).orElseThrow(
                    () -> new BadRequestException("Menu instance with id='" + menuId + "' doesn't exist")
            );
            newMenuList.add(new Menu(oldMenu.getItem()));
        }
        menuRp.deleteAll();
        menuRp.saveAll(newMenuList);

        return new ConfirmRes("Menu updated successfully", HttpStatus.OK);
    }
}

File: backend_nodejs_nest_fastify/my-pizza/src/nest_modules/product/services/menu.service.ts
import { BadRequestException, HttpStatus, Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Menu } from '../entities/menu.entity';
import { Repository } from 'typeorm';
import { paginate, Pagination, IPaginationOptions } from 'nestjs-typeorm-paginate';
import { UUID } from 'crypto';
import { ConfirmRes } from 'src/nest_modules/auth-user/interfaces/confirm-res.interface';
import { ProductService } from './product.service';
import { MenuItem } from '../entities/menu-item.entity';
import { MenuItemRes } from '../interfaces/menu-item-res.interface';
import { Product } from '../entities/product.entity';
import { Category } from '../entities/category.entity';
import { MenuRes } from '../interfaces/menu-res-interface';

@Injectable()
export class MenuService {

    constructor(@InjectRepository(Menu) private menuRepository: Repository<Menu>, private productSvc: ProductService) { }

    private generateMenuItemResModel(menuItem: MenuItem): MenuItemRes {
        if (menuItem instanceof Product) {
            return this.productSvc.generateProductResModel(menuItem)
        } else if (menuItem instanceof Category) {
            return this.productSvc.generateCategoryResModel(menuItem)
        }
    }


    public async getMenu(options: IPaginationOptions): Promise<Pagination<MenuRes>> {
        const results = await paginate<Menu>(this.menuRepository, options)
        return new Pagination<MenuRes>(
            results.items.map(menu => {
                return {
                    id: menu.id,
                    item: menu.item = this.generateMenuItemResModel(menu.item)
                }
            }),
            results.meta
        )
    }


    public async saveMenu(menuIds: UUID[]): Promise<ConfirmRes> {

        const newMenuList: Menu[] = []

        menuIds.forEach(async (menuId) => {
            const oldMenu: Menu | null | undefined = await this.menuRepository.findOneBy({ id: menuId })
            if (!oldMenu) throw new BadRequestException(`menu instance with id=${menuId} doesn't exist`,
                { cause: new Error(), description: 'Bad Request' })
            newMenuList.push(new Menu(oldMenu.item))
        })

        await this.menuRepository.createQueryBuilder('menu')
            .delete()
            .from(Menu)
            .execute()

        await this.menuRepository.save(newMenuList)

        return {
            statusCode: HttpStatus.NO_CONTENT,
            timestamp: new Date().getTime(),
            message: "Menu updated successfully"
        }

    }

}
